### ISSUE TITLE
Charting - real-time data update enhancements

### ISSUE THREAD

**Issue by dxbykov** [DevExpress/DevExtreme#5459]
# The Problem

The [Chart](https://js.devexpress.com/Demos/WidgetsGallery/Demo/Charts/Overview/) resets the zoom and scroll position when data is updated.

# The Proposed Solution

We are introducing the “visual range” concept (visible axis range). The user can move the visual range along the axis (scrolling) or change its length (zooming). You can define its behavior when the data source changes.

To set the visual range, specify its start and end values or length. The following code illustrates a visual range that equals the last 15 minutes on the axis:

```js
argumentAxis: { 
   visualRange: {
      length: {  
        minutes: 15 
      } 
  }
}
```

### Shift

When new data is added to the end, the visual range automatically moves to the axis’s end to include the new data. The range length remains unchanged.

```js
argumentAxis: {
  visualRangeUpdateMode: "shift" 
} 
```

### Keep

The visual range does not change.

This mode allows you to load data from the server partially. When the user zooms a specific range, pass the range to the server that should return a data subset for this range. Only the chart's data changes.

```js
argumentAxis: {
  visualRangeUpdateMode: "keep"
} 
```

### Reset

The visual range is reset to show all the chart’s data.

```js
argumentAxis: {
  visualRangeUpdateMode: "reset" 
} 
```

### Auto

This is the default mode. The visual range switches to a previously described mode depending on its position on the axis:

* At the start or middle, then `keep`;
* At the end, then `shift`;
* If set to view all data, then ‘reset’ (remains in the position to view all data)

This mode is optimized for data analysis: the visual range can be shifted to display the most recent data and does not move when you analyze trends or historical data.

```js
argumentAxis: {
  visualRangeUpdateMode: "auto" // the default value
} 
```

# We Need Your Feedback

## Take a Quick Poll

[How do the new Chart real-time data updates meet your needs?](https://docs.google.com/forms/d/e/1FAIpQLScqUT6_PUWC8Lpr0SmuimM-LphX2Qt87U_v4lq4eWxGVYtydA/viewform)

## Live Sandbox

[jQuery](https://codepen.io/anon/pen/qgXMZj)

## Installation

Follow the installation guide from the [release page](https://github.com/DevExpress/DevExtreme/releases) and leave us feedback on this topic. 

# Get Notified of Updates

Subscribe to this thread or our [Facebook](https://www.facebook.com/DevExpress.DevExtreme/) and [Twitter](https://twitter.com/devextreme) accounts for updates on this topic.

**Comment by dxbykov** [comment]
**Update:** An Angular example is available in the Live Sandbox section.

**Comment by dxbykov** [comment]
Thanks to everybody who gave us early feedback on this feature. The feature is now available in the [v18.2](https://github.com/DevExpress/DevExtreme/releases/tag/18.2.3) release. I'm closing this thread. In the case of bugs or questions, feel free to create new GitHub [issues](https://github.com/DevExpress/DevExtreme/issues) or tickets in our [Support Center](https://www.devexpress.com/Support/Center/).
